### PR TITLE
[ODS-6434]Rebuild Database Templates build -Skip already successful package creation from working branch

### PR DESCRIPTION
**Pull request by semalaiappan**
Below is a structured breakdown of each use case:

Use Case 1:
Scenario:

A new commit occurs in one of the three repositories (Ed-Fi-ODS, Ed-Fi-ODS-Implementation, or Ed-Fi-Extensions).
The workflow begins by updating the CodeGen package but fails before completing.
Expected Behavior:

In the first run, the workflow updates the CodeGen package version. However, the workflow fails after this update, preventing the completion of other package updates.
Example Workflow Run:

[Failed after updating CodeGen package](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/actions/runs/10517476403/job/29141657598).
Use Case 2:
Scenario:

The workflow is re-triggered after the failure in Use Case 1.
It should skip the CodeGen package update since it was already updated and continue updating the Extensions packages (TPDM, Sample, Homograph).
Expected Behavior:

The workflow should resume from the Extensions package update, bypassing the CodeGen package. However, it fails again after updating the Extensions packages.
Example Workflow Run:

[Failed after updating Extensions packages](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/actions/runs/10517520017/job/29141781325).
Use Case 3:
Scenario:

The workflow is re-triggered after the failure in Use Case 2.
It should skip both the CodeGen and Extensions package updates, resuming with updates to the EdFi.Ods.Minimal.Template, EdFi.Ods.Minimal.Template.PostgreSQL, EdFi.Ods.Populated.Template, and EdFi.Ods.Populated.Template.PostgreSQL packages.
Expected Behavior:

The workflow should skip the already updated CodeGen and Extensions packages, but it fails after attempting to update the Minimal and Populated templates.
Example Workflow Run:

[Failed after updating Minimal and Populated templates](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/actions/runs/10518959780/job/29145570132).
Use Case 4:
Scenario:

The workflow is re-triggered after the failure in Use Case 3.
It should skip the CodeGen, Extensions, and Minimal/Populated template updates and continue with the updates for EdFi.Ods.Minimal.Template.TPDM, EdFi.Ods.Minimal.Template.TPDM.PostgreSQL, EdFi.Ods.Populated.Template.TPDM, and EdFi.Ods.Populated.Template.TPDM.PostgreSQL packages.
Expected Behavior:

The workflow successfully completes by updating the TPDM-related templates without any failures.
Example Workflow Run:

[Successful completion after TPDM package updates](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/actions/runs/10519122224/job/29145994222).
Validation Steps:
Branch Consistency Across Repositories:
Ensure the working branch exists in all three repositories (Ed-Fi-ODS, Ed-Fi-ODS-Implementation, Ed-Fi-Extensions).
Check for Previous Workflow Failures:
Before starting the process, check if there was a failure in the last run. If so, determine whether any packages were updated successfully.
Handling New Commits:
If new commits are detected after a failed workflow, ensure the workflow resumes from where it left off by skipping already updated packages.

### CODE REVIEW

**Review by mjaksn — APPROVED** [review]

File: build.githubactions.ps1 [hunk 1]
```diff
@@ -288,6 +300,223 @@ function StandardVersions {
...
+        echo "$StatusEnvName=true" >> $Env:GITHUB_ENV
+    } else {
+        Write-Host "Package versions are the same:"
+        Write-Host "Branch ($branch): $branch_version"
+        Write-Host "Main: $main_version"
+        echo "$StatusEnvName=false" >> $Env:GITHUB_ENV
+    }
+}
+
+function AddOrUpdateRepositoriesJson {
```
Comment (vimayya) [inline build.githubactions.ps1:382]:
```suggestion
function CreateOrUpdateRepositoriesJson {
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 2]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+      run: |
+
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/branches/${{ env.current_branch }}"
+        .\build.githubactions.ps1  TestBranchExists -RepoName "Ed-Fi-ODS-Implementation" -BranchName "${{ env.current_branch }}" -Url $url
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/branches/${{ env.current_branch }}"
+        .\build.githubactions.ps1  TestBranchExists -RepoName "Ed-Fi-ODS" -BranchName "${{ env.current_branch }}" -Url $url
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-Extensions/branches/${{ env.current_branch }}"
+        .\build.githubactions.ps1  TestBranchExists -RepoName "Ed-Fi-Extensions" -BranchName "${{ env.current_branch }}" -Url $url
+
+    - name: Check any failed previous Workflow Run exist in current branch
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:114]:
```suggestion
    - name: Check for any failed previous workflow runs in the current branch
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 3]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+        workflow_conclusion: ("failure","cancelled")
+        path: ${{ github.workspace }}/Ed-Fi-ODS-Implementation/
+        run_id: ${{  env.rebuild_database_templates_lastrunId }}
+        repo: Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation
+        check_artifacts: true
+        allow_forks: false
+        if_no_artifact_found: ignore
+      continue-on-error: true
+
+    - name: Read each repo last commit for repo Ed-Fi-ODS-Implementation and stored it in repositories.json
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:138]:
```suggestion
    - name: Read the last commit for Ed-Fi-ODS-Implementation and store it in repositories.json 
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 4]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+      run: |
+
+        $oneeightyDaysAgo = (Get-Date).AddDays(-180).ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ")        
+        Write-Host "oneeightyDaysAgo Date and Time in ISO 8601 format: $oneeightyDaysAgo"
+        $branchName = "${{ env.current_branch }}"
+        $sinceDate = $oneeightyDaysAgo 
+        $implementationUrl = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/commits?sha=$branchName&since=$sinceDate&page=1&per_page=200"
+        .\build.githubactions.ps1 'AddOrUpdateRepositoriesJson' -Url $implementationUrl -RepoName 'Ed-Fi-ODS-Implementation'
+
+    - name: Read each repo last commit for repo Ed-Fi-ODS and stored it in repositories.json
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:150]:
```suggestion
    - name: Read the last commit for Ed-Fi-ODS and store it in repositories.json 
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 5]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+      run: |
+
+        $oneeightyDaysAgo = (Get-Date).AddDays(-180).ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ")
+        Write-Host "oneeightyDaysAgo Date and Time in ISO 8601 format: $oneeightyDaysAgo"
+        $branchName = "${{ env.current_branch }}"
+        $sinceDate = $oneeightyDaysAgo 
+        $odsUrl = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/commits?sha=$branchName&since=$sinceDate&page=1&per_page=200"
+        .\build.githubactions.ps1  'AddOrUpdateRepositoriesJson' -Url $odsUrl -RepoName 'Ed-Fi-ODS'
+
+    - name: Read each repo last commit for repo Ed-Fi-Extensions and stored it in repositories.json
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:162]:
```suggestion
    - name: Read the last commit for Ed-Fi-Extensions and store it in repositories.json 
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 6]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+        allow_forks: false
+        if_no_artifact_found: ignore
+      continue-on-error: true
+
+    - name: Read each repo last commit for repo Ed-Fi-ODS-Implementation and stored it in repositories.json 
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      shell: pwsh
+      run: |
+
+        $oneeightyDaysAgo = (Get-Date).AddDays(-180).ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ")
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:143]:
```suggestion
        $oneEightyDaysAgo= (Get-Date).AddDays(-180).ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ")        
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 7]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+        
+        # Read and parse the JSON file
+        $jsonContent = Get-Content -Path $FilePath | ConvertFrom-Json
+        
+        # Check if any repository has IscommitChanged set to true
+        $isCommitChanged = $jsonContent.repositories | Where-Object { $_.IscommitChanged -eq 'true' }
+
+        if ($isCommitChanged) {
+          # Set environment variable if any repository has IscommitChanged set to true
+          echo "IsSameCommitinAllthreeRepository=false" >> $Env:GITHUB_ENV
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:188]: can you change IsSameCommitinAllthreeRepository to NoNewCommitsInAllThreeRepositories for better readability.

File: .github/workflows/Rebuild Database Templates.yml [hunk 8]
```diff
@@ -88,8 +93,144 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.CodeGen.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/     
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.CodeGen.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'codeGen_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.CodeGen package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:225]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.CodeGen package version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 9]
```diff
@@ -147,15 +295,78 @@ jobs:
...
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+
+        $workflow_file_name = "Pkg EdFi.Ods.Extensions.Homograph.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-Extensions/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'Homograph_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Extensions.Homograph package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:329]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Extensions.Homograph package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 10]
```diff
@@ -147,15 +295,78 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Extensions.Sample.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Extensions.Sample.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-Extensions/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'Sample_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Extensions.Sample package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:345]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Extensions.Sample package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 11]
```diff
@@ -147,15 +295,78 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Extensions.TPDM.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Extensions.TPDM.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-Extensions/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'TPDM_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Extensions.TPDM package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:361]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Extensions.TPDM package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 12]
```diff
@@ -259,46 +470,151 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Populated.Template.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Populated.Template.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'populated_template_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Populated.Template package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:561]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Populated.Template package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 13]
```diff
@@ -259,46 +470,151 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Populated.Template.PostgreSQL.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Populated.Template.PostgreSQL.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'populated_template_postgreSQL_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Populated.Template.PostgreSQL package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:577]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Populated.Template.PostgreSQL package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 14]
```diff
@@ -259,46 +470,151 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Minimal.Template.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Minimal.Template.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'minimal_template_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Minimal.Template package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:593]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Minimal.Template package Version
```

File: .github/workflows/Rebuild Database Templates.yml [hunk 15]
```diff
@@ -259,46 +470,151 @@ jobs:
...
+    - name: Check if any successful workflow run in Pkg EdFi.Ods.Minimal.Template.PostgreSQL.yml
+      working-directory: ./Ed-Fi-ODS-Implementation/
+      if: ${{ env.IsSameCommitinAllthreeRepository == 'true' }}
+      shell: pwsh
+      run: |
+        $workflow_file_name = "Pkg EdFi.Ods.Minimal.Template.PostgreSQL.yml"
+        $url = "https://api.github.com/repos/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/actions/workflows/$workflow_file_name/runs?branch=${{ env.current_branch }}"
+        .\build.githubactions.ps1  'WorkflowCheck' -Url $url -ExpectedStatus 'completed' -ExpectedConclusions  @('success') -StatusEnvName 'minimal_template_postgreSQL_successful_runexist'
+  
+    - name: Compare already configuration.packages.json updated with EdFi.Ods.Minimal.Template.PostgreSQL package Version
```
Comment (vimayya) [inline .github/workflows/Rebuild Database Templates.yml:609]:
```suggestion
    - name: Check if configuration.packages.json has an updated EdFi.Ods.Minimal.Template.PostgreSQL package Version
```

**Comment by semalaiappan** [comment]
@vimayya  All review comments are fixed . Ready for Review again . Thanks

**Review by vimayya — APPROVED** [review]

**Review by mjaksn — APPROVED** [review]

**Pull request merged**